File: enable_layout.m
#import <Carbon/Carbon.h>
#import <Foundation/Foundation.h>

// ============================================================
// Mac-Arabic123-Fix Helper Tool (Compiled Objective-C Binary)
// Developed by Muhammad El-Shaikh
// GitHub:  github.com/ElShaikh1945
// Email:   Muhammad.Al-Shaikh@outlook.com
// Copyright © 2026 Muhammad El-Shaikh. All rights reserved.
// ============================================================

#define TARGET_ID @"org.sil.ukelele.keyboardlayout.arabic123pc.arabic-123-pc"

static NSString *getStrProp(TISInputSourceRef source, CFStringRef propertyKey) {
  CFStringRef val = (CFStringRef)TISGetInputSourceProperty(source, propertyKey);
  if (!val)
    return nil;
  return (__bridge NSString *)val;
}

static BOOL isEnabled(TISInputSourceRef source) {
  CFBooleanRef val = (CFBooleanRef)TISGetInputSourceProperty(
      source, kTISPropertyInputSourceIsEnabled);
  if (!val)
    return NO;
  return CFBooleanGetValue(val);
}

static BOOL isArabicKeyboard(TISInputSourceRef source) {
  NSString *category = getStrProp(source, kTISPropertyInputSourceCategory);
  if (![category
          isEqualToString:(__bridge NSString *)kTISCategoryKeyboardInputSource])
    return NO;

  NSString *sourceType = getStrProp(source, kTISPropertyInputSourceType);
  if (![sourceType isEqualToString:(__bridge NSString *)kTISTypeKeyboardLayout])
    return NO;

  NSString *sourceID = getStrProp(source, kTISPropertyInputSourceID);
  if (!sourceID)
    return NO;

  if ([sourceID isEqualToString:TARGET_ID] || [sourceID containsString:@"arabic123pc"])
    return NO;

  NSString *lowerID = [sourceID lowercaseString];
  if ([lowerID containsString:@"syriac"] || [lowerID containsString:@"mandaic"])
    return NO;

  if ([lowerID containsString:@"arabic"])
    return YES;

  NSString *name = [getStrProp(source, kTISPropertyLocalizedName) lowercaseString];
  if (name && ([name containsString:@"arabic"] || [name containsString:@"عربي"]))
    return YES;

  CFArrayRef langs = (CFArrayRef)TISGetInputSourceProperty(
      source, kTISPropertyInputSourceLanguages);
  if (langs) {
    NSArray *langArray = (__bridge NSArray *)langs;
    if ([langArray count] > 0) {
      id firstObj = [langArray firstObject];
      if ([firstObj isKindOfClass:[NSString class]]) {
        NSString *primary = (NSString *)firstObj;
        if ([primary isEqualToString:@"ar"] || [primary hasPrefix:@"ar-"] ||
            [primary hasPrefix:@"ar_"])
          return YES;
      }
    }
  }
  return NO;
}

static BOOL isTargetKeyboard(TISInputSourceRef src) {
  NSString *sid = getStrProp(src, kTISPropertyInputSourceID);
  NSString *name = getStrProp(src, kTISPropertyLocalizedName);
  return ([sid isEqualToString:TARGET_ID] ||
          [sid containsString:@"arabic123pc"] ||
          [name isEqualToString:@"Arabic - 123 - PC"] ||
          [name isEqualToString:@"عربي - 123 - PC"]);
}

static BOOL outputsDot(TISInputSourceRef source) {
  CFDataRef data = (CFDataRef)TISGetInputSourceProperty(source, kTISPropertyUnicodeKeyLayoutData);
  if (!data) return NO;
  const UCKeyboardLayout *layout = (const UCKeyboardLayout *)CFDataGetBytePtr(data);
  UInt32 keysDown = 0; UniChar chars[4]; UniCharCount len = 0;
  OSStatus status = UCKeyTranslate(layout, 65, kUCKeyActionDown, 0, 41, kUCKeyTranslateNoDeadKeysBit, &keysDown, 4, &len, chars);
  return (status == noErr && len > 0 && chars[0] == '.');
}

static void enableAndSelect() {
  NSString *homeBundle = [@"~/Library/Keyboard Layouts/Arabic - 123 - PC.bundle" stringByExpandingTildeInPath];
  NSString *sysBundle = @"/Library/Keyboard Layouts/Arabic - 123 - PC.bundle";
  NSFileManager *fm = [NSFileManager defaultManager];
  if ([fm fileExistsAtPath:homeBundle]) {
    TISRegisterInputSource((__bridge CFURLRef)[NSURL fileURLWithPath:homeBundle]);
  }
  if ([fm fileExistsAtPath:sysBundle]) {
    TISRegisterInputSource((__bridge CFURLRef)[NSURL fileURLWithPath:sysBundle]);
  }

  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources) {
    printf("Error: Could not retrieve input sources.\n");
    exit(1);
  }
  NSArray *list = (__bridge NSArray *)sources;
  TISInputSourceRef bestSource = NULL;
  NSMutableArray *outdatedSources = [NSMutableArray array];

  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    if (isTargetKeyboard(src)) {
      if (outputsDot(src)) {
        bestSource = src;
      } else {
        [outdatedSources addObject:(__bridge id)src];
      }
    }
  }

  if (!bestSource) {
    for (id obj in list) {
      TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
      if (isTargetKeyboard(src)) {
        bestSource = src;
        break;
      }
    }
  }

  for (id oldObj in outdatedSources) {
    TISInputSourceRef oldSrc = (__bridge TISInputSourceRef)oldObj;
    TISDisableInputSource(oldSrc);
  }

  if (bestSource) {
    TISEnableInputSource(bestSource);
    TISSelectInputSource(bestSource);
    printf("Successfully activated: Arabic - 123 - PC\n");
    fflush(stdout);
    CFRelease(sources);
    exit(0);
  } else {
    printf("Error: Custom layout not found in input sources list.\n");
    fflush(stdout);
    CFRelease(sources);
    exit(1);
  }
}

static void listArabic() {
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources)
    exit(0);
  NSArray *list = (__bridge NSArray *)sources;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    if (!isArabicKeyboard(src))
      continue;
    if (!isEnabled(src))
      continue;

    NSString *sid = getStrProp(src, kTISPropertyInputSourceID) ?: @"unknown";
    NSString *name = getStrProp(src, kTISPropertyLocalizedName) ?: sid;
    printf("%s|%s\n", [sid UTF8String], [name UTF8String]);
  }
  CFRelease(sources);
  exit(0);
}

static void disableIDs(NSString *idsStr) {
  NSArray *targetIDs = [idsStr componentsSeparatedByString:@","];
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources)
    exit(0);
  NSArray *list = (__bridge NSArray *)sources;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    NSString *sid = getStrProp(src, kTISPropertyInputSourceID);
    if (!sid)
      continue;
    for (NSString *tid in targetIDs) {
      NSString *cleanTid =
          [tid stringByTrimmingCharactersInSet:[NSCharacterSet
                                                   whitespaceCharacterSet]];
      if ([sid isEqualToString:cleanTid]) {
        TISDisableInputSource(src);
        break;
      }
    }
  }
  CFRelease(sources);
  exit(0);
}

static void checkInstalled() {
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources) {
    printf("NOT_INSTALLED\n");
    exit(1);
  }
  NSArray *list = (__bridge NSArray *)sources;
  BOOL found = NO;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    NSString *sid = getStrProp(src, kTISPropertyInputSourceID);
    NSString *name = getStrProp(src, kTISPropertyLocalizedName);
    if ([sid isEqualToString:TARGET_ID] ||
        [name isEqualToString:@"Arabic - 123 - PC"] ||
        [name isEqualToString:@"عربي - 123 - PC"]) {
      found = YES;
      BOOL en = isEnabled(src);
      printf("INSTALLED|%s|%s\n", [sid UTF8String],
             en ? "ENABLED" : "DISABLED");
      break;
    }
  }
  CFRelease(sources);
  if (found) {
    exit(0);
  } else {
    printf("NOT_INSTALLED\n");
    exit(1);
  }
}

static void cleanOldArabic() {
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources)
    exit(0);
  NSArray *list = (__bridge NSArray *)sources;
  int disabledCount = 0;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    if (!isArabicKeyboard(src))
      continue;
    if (!isEnabled(src))
      continue;

    NSString *name = getStrProp(src, kTISPropertyLocalizedName) ?: @"unknown";
    NSString *sid = getStrProp(src, kTISPropertyInputSourceID) ?: @"unknown";

    OSStatus status = TISDisableInputSource(src);
    if (status == noErr) {
      printf("Disabled: %s (%s)\n", [name UTF8String], [sid UTF8String]);
      disabledCount++;
    }
  }
  printf("Total disabled: %d\n", disabledCount);
  CFRelease(sources);
  exit(0);
}

static void listAllArabic() {
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources)
    exit(0);
  NSArray *list = (__bridge NSArray *)sources;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    if (!isArabicKeyboard(src))
      continue;

    NSString *sid = getStrProp(src, kTISPropertyInputSourceID) ?: @"unknown";
    NSString *name = getStrProp(src, kTISPropertyLocalizedName) ?: sid;
    BOOL en = isEnabled(src);
    printf("%s|%s|%s\n", [sid UTF8String], [name UTF8String],
           en ? "ENABLED" : "DISABLED");
  }
  CFRelease(sources);
  exit(0);
}

static void enableID(NSString *targetID) {
  NSString *cleanID = [targetID
      stringByTrimmingCharactersInSet:[NSCharacterSet whitespaceCharacterSet]];
  CFArrayRef sources = TISCreateInputSourceList(NULL, true);
  if (!sources) {
    printf("Error: Could not retrieve input sources.\n");
    exit(1);
  }
  NSArray *list = (__bridge NSArray *)sources;
  TISInputSourceRef found = NULL;
  for (id obj in list) {
    TISInputSourceRef src = (__bridge TISInputSourceRef)obj;
    NSString *sid = getStrProp(src, kTISPropertyInputSourceID);
    if ([sid isEqualToString:cleanID]) {
      found = src;
      break;
    }
  }
  if (found) {
    TISEnableInputSource(found);
    TISSelectInputSource(found);
    NSString *name = getStrProp(found, kTISPropertyLocalizedName) ?: cleanID;
    printf("Successfully activated: %s\n", [name UTF8String]);
    CFRelease(sources);
    exit(0);
  } else {
    printf("Error: Input source '%s' not found.\n", [cleanID UTF8String]);
    CFRelease(sources);
    exit(1);
  }
}

int main(int argc, const char *argv[]) {
  @autoreleasepool {
    setbuf(stdout, NULL);
    if (argc > 1) {
      NSString *arg1 = [NSString stringWithUTF8String:argv[1]];
      if ([arg1 isEqualToString:@"--check-installed"]) {
        checkInstalled();
      } else if ([arg1 isEqualToString:@"--list-arabic"]) {
        listArabic();
      } else if ([arg1 isEqualToString:@"--list-all-arabic"]) {
        listAllArabic();
      } else if ([arg1 isEqualToString:@"--enable-id"] && argc > 2) {
        NSString *arg2 = [NSString stringWithUTF8String:argv[2]];
        enableID(arg2);
      } else if ([arg1 isEqualToString:@"--disable-ids"] && argc > 2) {
        NSString *arg2 = [NSString stringWithUTF8String:argv[2]];
        disableIDs(arg2);
      } else if ([arg1 isEqualToString:@"--clean-old-arabic"]) {
        cleanOldArabic();
      } else {
        enableAndSelect();
      }
    } else {
      enableAndSelect();
    }
  }
  return 0;
}

File: enable_layout.swift
import Carbon
import Foundation

// ============================================================
// Mac-Arabic123-Fix v1.0.1
// Developed by Muhammad El-Shaikh
// GitHub:  github.com/ElShaikh1945
// Email:   Muhammad.Al-Shaikh@outlook.com
// Copyright © 2026 Muhammad El-Shaikh. All rights reserved.
// ============================================================

let targetID = "org.sil.ukelele.keyboardlayout.arabic123pc.arabic-123-pc"

// MARK: - Helper Functions

/// Get all input sources from macOS
func getAllInputSources() -> [TISInputSource] {
    guard let sources = TISCreateInputSourceList(nil, true)?.takeRetainedValue() as? [TISInputSource] else {
        return []
    }
    return sources
}

/// Get the string value of a TIS property
func getProperty(_ source: TISInputSource, _ key: CFString) -> String? {
    guard let cfValue = TISGetInputSourceProperty(source, key) else { return nil }
    return Unmanaged<AnyObject>.fromOpaque(cfValue).takeUnretainedValue() as? String
}

/// Check if an input source is an Arabic keyboard layout
func isArabicKeyboard(_ source: TISInputSource) -> Bool {
    guard let category = getProperty(source, kTISPropertyInputSourceCategory) else { return false }
    guard category == (kTISCategoryKeyboardInputSource as String) else { return false }
    
    guard let sourceType = getProperty(source, kTISPropertyInputSourceType) else { return false }
    guard sourceType == (kTISTypeKeyboardLayout as String) else { return false }
    
    guard let sourceID = getProperty(source, kTISPropertyInputSourceID) else { return false }
    if sourceID == targetID || sourceID.contains("arabic123pc") {
        return false
    }

    let lowerID = sourceID.lowercased()
    if lowerID.contains("syriac") || lowerID.contains("mandaic") {
        return false
    }

    if lowerID.contains("arabic") {
        return true
    }

    if let name = getProperty(source, kTISPropertyLocalizedName)?.lowercased(),
       name.contains("arabic") || name.contains("عربي") {
        return true
    }

    if let langsVal = TISGetInputSourceProperty(source, kTISPropertyInputSourceLanguages) {
        let langs = Unmanaged<AnyObject>.fromOpaque(langsVal).takeUnretainedValue() as? [String]
        if let primary = langs?.first {
            if primary == "ar" || primary.hasPrefix("ar-") || primary.hasPrefix("ar_") {
                return true
            }
        }
    }

    return false
}

/// Check if an input source is currently enabled
func isEnabled(_ source: TISInputSource) -> Bool {
    guard let cfValue = TISGetInputSourceProperty(source, kTISPropertyInputSourceIsEnabled) else { return false }
    return Unmanaged<AnyObject>.fromOpaque(cfValue).takeUnretainedValue() as? Bool ?? false
}

// MARK: - Command Handlers

/// Default: Enable and select the custom keyboard layout
func enableAndSelect() {
    let homeBundle = NSString(string: "~/Library/Keyboard Layouts/Arabic - 123 - PC.bundle").expandingTildeInPath
    let sysBundle = "/Library/Keyboard Layouts/Arabic - 123 - PC.bundle"
    if FileManager.default.fileExists(atPath: homeBundle) {
        TISRegisterInputSource(URL(fileURLWithPath: homeBundle) as CFURL)
    }
    if FileManager.default.fileExists(atPath: sysBundle) {
        TISRegisterInputSource(URL(fileURLWithPath: sysBundle) as CFURL)
    }

    let sources = getAllInputSources()
    
    var foundSource: TISInputSource? = nil
    for source in sources {
        let sid = getProperty(source, kTISPropertyInputSourceID) ?? ""
        let name = getProperty(source, kTISPropertyLocalizedName) ?? ""
        if sid == targetID || sid.contains("arabic123pc") || name == "Arabic - 123 - PC" || name == "عربي - 123 - PC" {
            foundSource = source
            break
        }
    }
    
    guard let source = foundSource else {
        print("Swift Error: Could not find keyboard layout '\(targetID)' in the installed layouts list.")
        print("Note: Please make sure the bundle has been copied correctly and the keyboard cache has refreshed.")
        exit(1)
    }
    
    print("Found custom keyboard layout: \(targetID)")
    
    // Enable the input source in the system settings
    let enableStatus = TISEnableInputSource(source)
    if enableStatus == noErr {
        print("Successfully enabled keyboard layout in System Settings.")
    } else {
        print("Warning: Failed to enable keyboard layout (Status Code: \(enableStatus)).")
    }
    
    // Select (activate) the input source as the active keyboard
    let selectStatus = TISSelectInputSource(source)
    if selectStatus == noErr {
        print("Successfully activated and switched to: Arabic - 123 - PC.")
    } else {
        print("Warning: Failed to switch to the new keyboard layout (Status Code: \(selectStatus)).")
    }
    
    exit(0)
}

/// List all enabled Arabic keyboard layouts (excluding our custom one)
func listArabic() {
    let sources = getAllInputSources()
    var found = false
    
    for source in sources {
        guard isArabicKeyboard(source) else { continue }
        guard isEnabled(source) else { continue }
        
        let sourceID = getProperty(source, kTISPropertyInputSourceID) ?? "unknown"
        let sourceName = getProperty(source, kTISPropertyLocalizedName) ?? sourceID
        
        print("\(sourceID)|\(sourceName)")
        found = true
    }
    
    if !found {
        // No Arabic keyboards found (other than ours), output nothing
        exit(0)
    }
    
    exit(0)
}

/// Disable specific input sources by their IDs (comma-separated)
func disableByIDs(_ idsString: String) {
    let idsToDisable = idsString.components(separatedBy: ",").map { $0.trimmingCharacters(in: .whitespaces) }
    let sources = getAllInputSources()
    var disabledCount = 0
    
    for source in sources {
        guard let sourceID = getProperty(source, kTISPropertyInputSourceID) else { continue }
        
        if idsToDisable.contains(sourceID) {
            let status = TISDisableInputSource(source)
            if status == noErr {
                let name = getProperty(source, kTISPropertyLocalizedName) ?? sourceID
                print("Disabled: \(name) (\(sourceID))")
                disabledCount += 1
            } else {
                print("Warning: Failed to disable \(sourceID) (Status Code: \(status)).")
            }
        }
    }
    
    print("Total disabled: \(disabledCount)")
    exit(0)
}

/// Disable all Arabic keyboards except our custom one
func cleanOldArabic() {
    let sources = getAllInputSources()
    var disabledCount = 0
    
    for source in sources {
        guard isArabicKeyboard(source) else { continue }
        guard isEnabled(source) else { continue }
        
        let sourceID = getProperty(source, kTISPropertyInputSourceID) ?? "unknown"
        let name = getProperty(source, kTISPropertyLocalizedName) ?? sourceID
        
        let status = TISDisableInputSource(source)
        if status == noErr {
            print("Disabled: \(name) (\(sourceID))")
            disabledCount += 1
        } else {
            print("Warning: Failed to disable \(sourceID) (Status Code: \(status)).")
        }
    }
    
    print("Total disabled: \(disabledCount)")
    exit(0)
}

/// Check if the custom keyboard layout is already installed/present in the system
func checkInstalled() {
    let sources = getAllInputSources()
    for source in sources {
        let sid = getProperty(source, kTISPropertyInputSourceID) ?? ""
        let name = getProperty(source, kTISPropertyLocalizedName) ?? ""
        if sid == targetID || name == "Arabic - 123 - PC" || name == "عربي - 123 - PC" {
            let enabled = isEnabled(source)
            print("INSTALLED|\(sid)|\(enabled ? "ENABLED" : "DISABLED")")
            exit(0)
        }
    }
    print("NOT_INSTALLED")
    exit(1)
}

/// List all installed Arabic keyboard layouts (enabled or disabled)
func listAllArabic() {
    let sources = getAllInputSources()
    for source in sources {
        guard isArabicKeyboard(source) else { continue }
        let sourceID = getProperty(source, kTISPropertyInputSourceID) ?? "unknown"
        let sourceName = getProperty(source, kTISPropertyLocalizedName) ?? sourceID
        let en = isEnabled(source)
        print("\(sourceID)|\(sourceName)|\(en ? "ENABLED" : "DISABLED")")
    }
    exit(0)
}

/// Enable and select a specific input source by ID
func enableByID(_ targetIDToEnable: String) {
    let cleanID = targetIDToEnable.trimmingCharacters(in: .whitespaces)
    let sources = getAllInputSources()
    for source in sources {
        guard let sourceID = getProperty(source, kTISPropertyInputSourceID), sourceID == cleanID else { continue }
        TISEnableInputSource(source)
        TISSelectInputSource(source)
        let name = getProperty(source, kTISPropertyLocalizedName) ?? cleanID
        print("Successfully activated: \(name)")
        exit(0)
    }
    print("Error: Input source '\(cleanID)' not found.")
    exit(1)
}

// MARK: - Main Entry Point

let args = CommandLine.arguments

if args.count > 1 {
    switch args[1] {
    case "--check-installed":
        checkInstalled()
    case "--list-arabic":
        listArabic()
    case "--list-all-arabic":
        listAllArabic()
    case "--enable-id":
        guard args.count > 2 else {
            print("Error: --enable-id requires an input source ID.")
            exit(1)
        }
        enableByID(args[2])
    case "--disable-ids":
        guard args.count > 2 else {
            print("Error: --disable-ids requires a comma-separated list of input source IDs.")
            print("Usage: enable_layout --disable-ids \"com.apple.keylayout.Arabic,com.apple.keylayout.Arabic-PC\"")
            exit(1)
        }
        disableByIDs(args[2])
    case "--clean-old-arabic":
        cleanOldArabic()
    default:
        print("Unknown option: \(args[1])")
        print("Usage:")
        print("  enable_layout                          Enable and select Arabic-123-PC")
        print("  enable_layout --check-installed        Check if Arabic-123-PC is installed")
        print("  enable_layout --list-arabic            List enabled Arabic keyboards")
        print("  enable_layout --list-all-arabic        List all installed Arabic keyboards")
        print("  enable_layout --enable-id \"ID\"         Enable and select a keyboard by ID")
        print("  enable_layout --disable-ids \"ID1,ID2\"  Disable specific keyboards by ID")
        print("  enable_layout --clean-old-arabic       Disable all Arabic keyboards except Arabic-123-PC")
        exit(1)
    }
} else {
    enableAndSelect()
}
